### PR TITLE
Add execute function access control

### PR DESCRIPTION
**Pull request by kokosing**
Add execute function access control

### CODE REVIEW

**Comment by kokosing** [comment]
It is still very WIP

**Comment by kokosing** [comment]
It is ready to be reviewed.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java [hunk 1]
```diff
@@ -1109,6 +1111,52 @@ public void testViewColumnAccessControl()
...
+    {
+        skipTestUnless(supportsViews());
+
+        Session viewOwnerSession = TestingSession.testSessionBuilder()
+                .setIdentity(Identity.ofUser("test_view_access_owner"))
+                .setCatalog(getSession().getCatalog().get())
+                .setSchema(getSession().getSchema().get())
+                .build();
+
+        // TEST FUNCTION PRIVILEGES
```
Comment (skrzypo987) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java]: Instead of comments you can extract those to different methods

File: presto-main/src/test/java/io/prestosql/operator/scalar/FunctionAssertions.java [hunk 2]
```diff
@@ -349,7 +347,7 @@ public void assertCachedInstanceHasBoundedRetainedSize(String projection)
     {
         requireNonNull(projection, "projection is null");
 
-        Expression projectionExpression = createExpression(session, projection, metadata, TypeProvider.copyOf(INPUT_TYPES));
+        Expression projectionExpression = ExpressionTestUtils.createExpression(session, projection, metadata, TypeProvider.copyOf(INPUT_TYPES));
```
Comment (skrzypo987) [inline presto-main/src/test/java/io/prestosql/operator/scalar/FunctionAssertions.java]: static import would be nice here

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java (hunk 1 above)
Comment (kokosing) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java]: I think these tests requires more love than that: https://github.com/prestosql/presto/issues/3037

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java [hunk 3]
```diff
@@ -1070,43 +1072,89 @@ public void testViewAccessControl()
...
-        assertAccessAllowed(nestedViewOwnerSession, "DROP VIEW test_nested_view_access");
-        assertAccessAllowed(viewOwnerSession, "DROP VIEW test_view_access");
-        assertAccessAllowed(viewOwnerSession, "DROP VIEW test_invoker_view_access");
+        assertAccessAllowed(nestedViewOwnerSession, "DROP VIEW test_nested_view_column_access");
+        assertAccessAllowed(viewOwnerSession, "DROP VIEW test_view_column_access");
+        assertAccessAllowed(viewOwnerSession, "DROP VIEW test_invoker_view_column_access");
+    }
+
+    @Test
+    public void testViewFunctionAccessControl()
```
Comment (lukasz-walkiewicz) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java]: That's too many test scenarios in single method. It's hard to follow what is actually happening.

Comment (kokosing) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java]: I would like to refactor these tests in: #3037

Currently it follows same pattern as `testViewColumnAccessControl`

File: presto-main/src/main/java/io/prestosql/security/AccessControl.java [hunk 4]
```diff
@@ -255,6 +255,13 @@
      */
     void checkCanCreateViewWithSelectFromColumns(SecurityContext context, QualifiedObjectName tableName, Set<String> columnNames);
 
+    /**
+     * Check if identity is allowed to create a view that executes the function.
+     *
+     * @throws io.prestosql.spi.security.AccessDeniedException if not allowed
+     */
+    void checkCanCreateViewWithExecuteFunction(SecurityContext context, String functionName);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/security/AccessControl.java]: I think this is the wrong abstraction. There's nothing in the SQL spec that limits whether user can create a view based on whether it contains certain functions. All it says is that:

```
A column privilege descriptor is created that defines the privilege SELECT on CV to A. 
That privilege is grantable if and only if all the following are true:
...
2) The applicable privileges for A include grantable EXECUTE privileges on all SQL-invoked 
routines that are subject routines of <routine invocation>s contained in QE.
...
```

which, basically, states that the creator of the view (i.e., the schema owner) has a SELECT privilege on each column of the view. That privilege is grantable to others if certain conditions apply (i.e., among other things, the creator has a grantable EXECUTE privilege on all functions referenced by the query)

File: presto-main/src/main/java/io/prestosql/security/AccessControlManager.java [hunk 5]
```diff
@@ -24,6 +24,7 @@
 import io.prestosql.metadata.QualifiedObjectName;
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/security/AccessControlManager.java]: I'd rephrase the commit as "Improve robustness of InitializeSystemAccessControl", and in the commit message indicate that this makes it easier to evolve the SystemAccessControl interface by adding new methods.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestEffectivePredicateExtractor.java [hunk 6]
```diff
@@ -607,6 +601,23 @@ public void testValues()
...
+        return transaction(new TestingTransactionManager(), new AllowAllAccessControl())
+                .singleStatement()
+                .execute(SESSION, transactionSession -> {
+                    return effectivePredicateExtractor.extract(
+                            transactionSession,
+                            new ValuesNode(
+                                    newId(),
+                                    ImmutableList.of(A, B),
+                                    ImmutableList.of(
+                                            ImmutableList.of(bigintLiteral(1), new FunctionCall(resolvedFunction.toQualifiedName(), ImmutableList.of())))),
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestEffectivePredicateExtractor.java]: This doesn't seem to belong to this "generic" method.

File: presto-benchmark/src/main/java/io/prestosql/benchmark/AbstractOperatorBenchmark.java [hunk 7]
```diff
@@ -279,6 +279,7 @@ protected final OperatorFactory createHashProjectOperator(int operatorId, PlanNo
     {
```
Comment (martint) [inline presto-benchmark/src/main/java/io/prestosql/benchmark/AbstractOperatorBenchmark.java]: Can you add a comment in the commit message describing the motivation for this change?

File: presto-main/src/main/java/io/prestosql/security/AccessControlManager.java [hunk 8]
```diff
@@ -586,6 +586,15 @@ public void checkCanCreateViewWithSelectFromColumns(SecurityContext securityCont
         catalogAuthorizationCheck(tableName.getCatalogName(), securityContext, (control, context) -> control.checkCanCreateViewWithSelectFromColumns(context, tableName.asSchemaTableName(), columnNames));
     }
 
+    @Override
+    public void checkCanCreateViewWithExecuteFunction(SecurityContext securityContext, String functionName)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/security/AccessControlManager.java]: Change name as discussed offline

File: presto-main/src/main/java/io/prestosql/sql/planner/ExpressionInterpreter.java [hunk 9]
```diff
@@ -174,7 +175,7 @@ public static ExpressionInterpreter expressionOptimizer(Expression expression, M
 
     public static Object evaluateConstantExpression(Expression expression, Type expectedType, Metadata metadata, Session session, Map<NodeRef<Parameter>, Expression> parameters)
     {
-        ExpressionAnalyzer analyzer = createConstantAnalyzer(metadata, session, parameters, WarningCollector.NOOP);
+        ExpressionAnalyzer analyzer = createConstantAnalyzer(metadata, new AllowAllAccessControl(), session, parameters, WarningCollector.NOOP);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/ExpressionInterpreter.java]: This method is used in various places, such as when resolving session properties, arguments of procedures, etc. It should use a real AccessControl.

File: presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java [hunk 10]
```diff
@@ -1070,43 +1072,89 @@ public void testViewAccessControl()
...
+        // TEST FUNCTION PRIVILEGES
+        // view creation permissions are only checked at query time, not at creation
+        assertAccessAllowed(
+                viewOwnerSession,
+                "CREATE VIEW test_view_function_access AS SELECT abs(1) AS c",
+                privilege("abs", GRANT_EXECUTE_FUNCTION));
+
+        assertAccessDenied(
+                "SELECT * FROM test_view_function_access",
+                "'test_view_access_owner' cannot grant 'abs' execution to user 'hive'",
```
Comment (kokosing) [inline presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java]: This is a bit misleading error. Do you have any suggestions?

Comment (martint) [reply presto-testing/src/main/java/io/prestosql/testing/AbstractTestDistributedQueries.java]: That's good enough. Eventually we should improve these messages by providing the whole chain of view references that lead to such errors.
